### ISSUE TITLE
Not able to create device context for client in wfreerdp-server.

### ISSUE THREAD

**Issue by dashingvik** [FreeRDP/FreeRDP#5716]
In wf_mirage.c -> createDC function returns NULL where wfi->deviceName = \\.\DISPLAYV1
Check with GetLastError() which returns 0 means no error.

![image](https://user-images.githubusercontent.com/56148491/68662424-c77ccc00-0562-11ea-852a-c9efdaaf1c12.png)

but when I change the value to "DISPLAY" it gives me following values but null check is bypassed as of now (which gives me error in future code discussed at end).

![image](https://user-images.githubusercontent.com/56148491/68663309-6f46c980-0564-11ea-85aa-8a1c955499f1.png)

- I have installed Mirage Driver and using it is as mirror driver.
- I am not sure if this is a bug or problem at my side (if yes, what's that).
- [Similar to] ContextNew callback failed. Failed to create new peer context #4675
- There is no problem with security like firewall, defender or an antivirus (all are disabled).
- On client side I am able to see only a black screen with no error as follows :
![image](https://user-images.githubusercontent.com/56148491/68663982-8d60f980-0565-11ea-9828-a9c856012b98.png)

- Here's the error that i got and not able to solved this as this referenced to creaeDC()
![image](https://user-images.githubusercontent.com/56148491/68664054-b3869980-0565-11ea-8b04-bf9f8cd08ded.png)

- Commented a bit of code to see if this is working or not which is as below because it always returns me '0'.
![image](https://user-images.githubusercontent.com/56148491/68664357-48899280-0566-11ea-82a2-d521a8f6696e.png)

**Comment by akallabeth** [comment]
@dashingvik sorry, can´t help much there. It has been broken for some time but not much has been fixed (and currently mostly linux/mac devs are providing pr)
That aside, happy to help out on IRC or the mailing list as well as we can, but here you need to debug yourself :/

**Comment by dashingvik** [comment]
@akallabeth Is it working your side ? If yes, can you let me know the value of following variable in wf_mirage.c so that I can do some R&D ?

- wfi->changeBuffer & status in [ ExtEscape(wfi->driverDC, dmf_esc_usm_pipe_map, 0, 0, sizeof(GETCHANGESBUF), (LPSTR) wfi->changeBuffer); ]
-wfi->deviceName in CreateDC();

**Comment by akallabeth** [comment]
@dashingvik no, did not test it, we currently have not many people reporting/fixing windows side bugs.

**Comment by dashingvik** [comment]
@akallabeth Is there any one who can help me with this issue as I am struggling here for so many days and I am not able to find any work around :/

**Comment by sfhacker** [comment]
@dashingvik Is this open issue resolved? if not, we might be able to assist. We are just users but we will try to reproduce this issue the best we can. There are still a lot of outstanding issues since last year that no one seems to know how to fix them. Please post here or email us how to reproduce this issue and we will take it from there. Cheers.

**Comment by dashingvik** [comment]
> @dashingvik Is this open issue resolved? if not, we might be able to assist. We are just users but we will try to reproduce this issue the best we can. There are still a lot of outstanding issues since last year that no one seems to know how to fix them. Please post here or email us how to reproduce this issue and we will take it from there. Cheers.

@sfhacker Hey, your reply brings back some hope for me. I am still stuck at that issue and finding some work around for it. I am providing my email ID so that we can be in touched for further communication :)

Email-ID => vivek.sonawane@sectona.com

**Comment by zhangleia** [comment]
The problem of not connecting to wfreerdp server in Windows 7 system,

The solution I found is to first install the VNC dfmirage-setup-2.0.301 driver, compile and install the service segment of tightvnc. After using the client connection, and after tightvnc initializes and activates the driver, win7's wfreerdp server can be connected successfully

**Comment by zhangleia** [comment]
I tested and changed the wf_mirage.c file to realize the connection of the windows 7 system

BOOL wf_mirror_driver_update(wfInfo* wfi, int mode)
{
	HDC dc;
	BOOL status;
	DWORD* extHdr;
	WORD drvExtraSaved;
	DEVMODE* deviceMode;
	LONG disp_change_status;
	DWORD dmf_devmodewext_magic_sig = 0xDF20C0DE;

	if ((mode != MIRROR_LOAD) && (mode != MIRROR_UNLOAD))
	{
		WLog_DBG(TAG, "Invalid mirror mode!");
		return FALSE;
	}

	deviceMode = (DEVMODE*)malloc(sizeof(DEVMODE) + EXT_DEVMODE_SIZE_MAX);

	if (!deviceMode)
		return FALSE;

	deviceMode->dmDriverExtra = 2 * sizeof(DWORD);
	extHdr = (DWORD*)((BYTE*)&deviceMode + sizeof(DEVMODE));
	extHdr[0] = dmf_devmodewext_magic_sig;
	extHdr[1] = 0;
	drvExtraSaved = deviceMode->dmDriverExtra;
	memset(deviceMode, 0, sizeof(DEVMODE) + EXT_DEVMODE_SIZE_MAX);
	deviceMode->dmSize = sizeof(DEVMODE);
	deviceMode->dmDriverExtra = drvExtraSaved;

	if (mode == MIRROR_LOAD)
	{
		wfi->virtscreen_width = GetSystemMetrics(SM_CXVIRTUALSCREEN);
		wfi->virtscreen_height = GetSystemMetrics(SM_CYVIRTUALSCREEN);

		EnumDisplaySettings(NULL, ENUM_CURRENT_SETTINGS, deviceMode);
		/*
		deviceMode->dmPelsWidth = wfi->virtscreen_width;
		deviceMode->dmPelsHeight = wfi->virtscreen_height;
		deviceMode->dmBitsPerPel = wfi->bitsPerPixel;
		deviceMode->dmPosition.x = wfi->servscreen_xoffset;
		deviceMode->dmPosition.y = wfi->servscreen_yoffset;
		*/
	}

	deviceMode->dmFields = DM_BITSPERPEL | DM_PELSWIDTH | DM_PELSHEIGHT | DM_POSITION;
	_tcsncpy_s(deviceMode->dmDeviceName, 32, wfi->deviceName, _tcslen(wfi->deviceName));
	disp_change_status =
	    ChangeDisplaySettingsEx(wfi->deviceName, deviceMode, NULL, CDS_UPDATEREGISTRY, NULL);
	status = (disp_change_status == DISP_CHANGE_SUCCESSFUL) ? TRUE : FALSE;

	if (!status)
		wf_mirror_driver_print_display_change_status(disp_change_status);

	return status;
}

**Comment by sonyps5201314** [comment]
It seems that the reason is  `bitsPerPixel==0`
and apply the follow diff code:
`
extHdr = (DWORD*)((BYTE*)&deviceMode + sizeof(DEVMODE));
`
change to 
`
extHdr = (DWORD*)((BYTE*)deviceMode + sizeof(DEVMODE));
`
can prevents stack overwrite when the client exits.

**Comment by sfhacker** [comment]
This open issue is from 2019! Wow!
